File: app/src/androidTest/java/social/entourage/android/LoginTest.java
package social.entourage.android;

import android.content.Context;
import android.view.autofill.AutofillManager;

import androidx.test.espresso.IdlingRegistry;
import androidx.test.espresso.IdlingResource;
import androidx.test.filters.LargeTest;
import androidx.test.rule.ActivityTestRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.jakewharton.espresso.OkHttp3IdlingResource;

import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import okhttp3.OkHttpClient;
import social.entourage.android.authentication.login.LoginActivity;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.closeSoftKeyboard;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.doesNotExist;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class LoginTest {
    @Rule
    public ActivityTestRule<LoginActivity> activityRule = new ActivityTestRule<>(LoginActivity.class);

    private IdlingResource resource;
    private AutofillManager afM;

    @Before
    public void setUp() {
        checkNoUserIsLoggedIn();
        Context context = activityRule.getActivity();
        OkHttpClient client = EntourageApplication.get(context).getEntourageComponent().getOkHttpClient();
        resource = OkHttp3IdlingResource.create("OkHttp", client);
        IdlingRegistry.getInstance().register(resource);
        afM = context.getSystemService(AutofillManager.class);
        if(afM!=null) {
            afM.disableAutofillServices();
        }
    }

    private void checkTCDisplay() {
        //PFP we have a TC validation screen
        if(EntourageApplication.isPfpApp()) {
            onView(withId(R.id.register_welcome_start_button)).perform(click());
        }
    }

    private void checkNoUserIsLoggedIn() {
        Context context = activityRule.getActivity();
        if(EntourageApplication.get(context).getEntourageComponent().getAuthenticationController().getUser() != null) {
           EntourageApplication.get(context).getEntourageComponent().getAuthenticationController().logOutUser();
        }
    }

    @After
    public void tearDown() {
        IdlingRegistry.getInstance().unregister(resource);
        checkNoUserIsLoggedIn();
    }

    @Test
    public void loginOK() {
        checkNoUserIsLoggedIn();
        onView(withId(R.id.login_button_login)).perform(click());
        checkTCDisplay();

        onView(withId(R.id.login_edit_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_edit_code)).perform(typeText(BuildConfig.TEST_ACCOUNT_PWD), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_button_signin)).perform(click());

        onView(withText(R.string.login_error_title)).check(doesNotExist());
        //checkNoUserIsLoggedIn();
    }

    @Test
    public void loginOKwithoutCountryCode() {
        onView(withId(R.id.login_button_login)).perform(click());
        checkTCDisplay();

        onView(withId(R.id.login_edit_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN.replaceFirst("\\+33", "0")), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_edit_code)).perform(typeText(BuildConfig.TEST_ACCOUNT_PWD), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_button_signin)).perform(click());

        onView(withText(R.string.login_error_title)).check(doesNotExist());
        //checkNoUserIsLoggedIn();
    }

    private void closeAutofill() {
        if(afM ==null) {
            afM = activityRule.getActivity().getSystemService(AutofillManager.class);
        }
        if(afM!=null) {
            afM.cancel();
            afM.commit();
        }
    }

    @Test
    public void loginFailureWrongPassword() {
        checkNoUserIsLoggedIn();
        onView(withId(R.id.login_button_login)).perform(click());
        checkTCDisplay();

        onView(withId(R.id.login_edit_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_edit_code)).perform(typeText("999999"), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_button_signin)).perform(click());

        onView(withText(R.string.login_error_title)).check(matches(isDisplayed()));
        onView(withText(R.string.login_retry_label)).perform(click());
        onView(withId(R.id.login_back_button)).perform(click());
    }

    @Test
    public void loginFailureWrongPhoneNumberFormat() {
        onView(withId(R.id.login_button_login)).perform(click());
        checkTCDisplay();

        onView(withId(R.id.login_edit_phone)).perform(typeText("012345678"), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_edit_code)).perform(typeText("000000"), closeSoftKeyboard());
        closeAutofill();
        onView(withId(R.id.login_button_signin)).perform(click());

        onView(withText(R.string.login_error_title)).check(matches(isDisplayed()));
        onView(withText(R.string.login_retry_label)).perform(click());
        onView(withId(R.id.login_back_button)).perform(click());
    }
}

File: app/src/androidTest/java/social/entourage/android/LoginTest.kt
package social.entourage.android

import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.clearText
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.closeSoftKeyboard
import androidx.test.espresso.action.ViewActions.typeText
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.intent.Intents.intended
import androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.filters.LargeTest
import org.hamcrest.Matchers.allOf
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import social.entourage.android.onboarding.login.LoginActivity
import social.entourage.android.onboarding.login.LoginChangePhoneActivity
import social.entourage.android.onboarding.pre_onboarding.PreOnboardingChoiceActivity

@RunWith(AndroidJUnit4::class)
@LargeTest
class LoginTest : EntourageTestBeforeLogin() {
    @get:Rule
    val activityRule = ActivityScenarioRule(LoginActivity::class.java)

    @Before
    fun setUp() {
        activityRule.scenario.onActivity { activity ->
            super.setUp(activity)
        }
        checkFirstConnectionScreen()
    }

    @After
    override fun tearDown() {
        //keep it just for the annotation
        super.tearDown()
    }
    private fun closeAutofill() {
        activityRule.scenario.onActivity { activity ->
            closeAutofill(activity)
        }
    }

    @Test
    fun loginOK() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText(BuildConfig.TEST_ACCOUNT_PWD), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())
        checkLoginSuccessful()
    }

    @Test
    fun loginOKWithoutCountryCode() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN.replaceFirst("\\+33".toRegex(), "0")), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText(BuildConfig.TEST_ACCOUNT_PWD), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())
        checkLoginSuccessful()
    }

    @Test
    fun loginFailureWrongPassword() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText("999999"), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())
        checkLoginFailure()
    }

    @Test
    fun loginFailureShortPassword() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText("9999"), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())
        checkLoginFailure(R.string.attention_pop_title, R.string.close)
    }

    @Test
    fun loginFailureWrongPhoneNumberFormat() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText("012345678"), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText("000000"), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())

        checkLoginFailure()
    }

    @Test
    fun loginFailureNoInternetConnection() {
        //Disable wifi and data
        enableWifiAndData(false)

        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_et_code)).perform(typeText(BuildConfig.TEST_ACCOUNT_PWD), closeSoftKeyboard())
        closeAutofill()
        onView(withId(R.id.ui_login_button_signup)).perform(click())

        //Check that error is displayed
        onView(withText(R.string.login_error_network)).check(matches(isDisplayed()))
        //onView(withText(R.string.login_retry_label)).perform(click())
    }

    @Test
    fun resendCodeButtonWithoutClick() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())

        onView(withId(R.id.ui_login_button_resend_code)).perform(click())

        onView(withText(R.string.login_button_resend_code_action)).check(matches(isDisplayed()))
    }

    @Test
    fun resendCodeButtonWithEmptyPhoneNumber() {
        onView(withId(R.id.ui_login_phone_et_phone)).perform(clearText())

        onView(withId(R.id.ui_login_button_resend_code)).perform(click())

        checkLoginFailure(R.string.attention_pop_title, R.string.close)
    }

    @Test
    fun displayChangePhoneNumberScreen() {
        onView(withId(R.id.ui_login_button_change_phone)).perform(click())
        intended(hasComponent(LoginChangePhoneActivity::class.java.name))
    }

    @Test
    fun clickGoBack() {
        onView(withId(R.id.icon_back)).perform(click())
        intended(hasComponent(PreOnboardingChoiceActivity::class.java.name))
    }

    @Test
    fun displayTermsAndConditions() {
        onView(withId(R.id.tv_condition_generales))
            .perform(click())
        //TODO intended(hasComponent(PreOnboardingChoiceActivity::class.java.name))
    }

    @Test
    fun resendCodeFailureNoInternetConnection() {
        //Disable wifi and data
        enableWifiAndData(false)

        //Try to resend code
        onView(withId(R.id.ui_login_phone_et_phone)).perform(typeText(BuildConfig.TEST_ACCOUNT_LOGIN), closeSoftKeyboard())
        closeAutofill()
        onView(allOf(withId(R.id.ui_login_button_resend_code), isDisplayed())).perform(click())
        onView(allOf(withText(R.string.login_button_resend_code_action), isDisplayed())).perform(click())

        //Check that error is displayed
        //TODO onView(withText(R.string.login_error_network)).inRoot(ToastMatcher()).check(matches(isDisplayed()))

    }

    /*class ToastMatcher : TypeSafeMatcher<Root>() {
        override fun matchesSafely(item: Root?): Boolean {
            item?.windowLayoutParams?.get()?.type?.let { type ->
                if (type == WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY) {
                    val windowToken = item.decorView.windowToken
                    val appToken = item.decorView.applicationWindowToken
                    if (windowToken == appToken) {
                        //Means this window isn't contained by any other windows
                        return true
                    }
                }
            }
            return false
        }

        override fun describeTo(description: Description?) {}
    }*/
}